Array based clUpdateMutableCommandsKHR changes

EwanC opened the pull request:
CTS changes to reflect the spec changes merged in https://github.com/KhronosGroup/OpenCL-Docs/pull/1045 and requires header updates from https://github.com/KhronosGroup/OpenCL-Headers/pull/245

Tested using OCK implementation from https://github.com/codeplaysoftware/oneapi-construction-kit/pull/501

bashbaug reviewed (commented):
I don't have an easy way to test these changes (unless I pull and build the latest OCK) but they all look reasonable.

I think it's a good proof-point that the versioned extension headers work, but for the CTS tests I think we've generally been OK tracking the latest headers.  Since we don't regularly test (or even build) with older headers I think the older code would eventually break anyhow.

I think it would be a good idea to skip these tests for older versions of the extension though, otherwise they'll likely crash unpredictably if the newer functions run on an older implementation.  Could this be added?

EwanC commented:
> I don't have an easy way to test these changes (unless I pull and build the latest OCK) but they all look reasonable.
> 
> I think it's a good proof-point that the versioned extension headers work, but for the CTS tests I think we've generally been OK tracking the latest headers. Since we don't regularly test (or even build) with older headers I think the older code would eventually break anyhow.
> 
> I think it would be a good idea to skip these tests for older versions of the extension though, otherwise they'll likely crash unpredictably if the newer functions run on an older implementation. Could this be added?

I've removed the #ifdefs so this PR will require the matching headers to work. Also added in a [CL_DEVICE_EXTENSIONS_WITH_VERSION](https://registry.khronos.org/OpenCL/specs/3.0-unified/html/OpenCL_API.html#CL_DEVICE_EXTENSIONS_WITH_VERSION) runtime check and skipped testing when a device doesn't support at least 0.9.2 of mutable-dispatch. Although devices which don't support that query will still run into issues

bashbaug commented:
Merging as discussed in the September 3rd teleconference + email.
